### ISSUE TITLE
Indexing error: "TypeError: Table and inner RecordBatch schemas must be equivalent."

### ISSUE THREAD

**Issue by selfagency** [continuedev/continue#2026]
### Before submitting your bug report

- [X] I believe this is a bug. I'll try to join the [Continue Discord](https://discord.gg/NWtdYexhMs) for questions
- [X] I'm not able to find an [open issue](https://github.com/continuedev/continue/issues?q=is%3Aopen+is%3Aissue) that reports the same bug
- [X] I've seen the [troubleshooting guide](https://docs.continue.dev/troubleshooting) on the Continue Docs

### Relevant environment info

```Markdown
- OS: Ubuntu 20 on WSL via Windows 11 Pro
- Continue: v0.8.46
- IDE: VS Code
- Model: Claude 3.5 Sonnet or GPT-4o
- config.json:
  
{
  "models": [
    {
      "model": "claude-3-5-sonnet-20240620",
      "provider": "anthropic",
      "apiKey": "Redacted",
      "title": "Claude 3.5 Sonnet"
    },
    {
      "model": "gpt-4o",
      "contextLength": 128000,
      "apiKey": "REDACTED",
      "title": "GPT-4o",
      "systemMessage": "You are an expert software developer. You give helpful and concise responses.",
      "provider": "openai"
    }
  ],
  "customCommands": [
    {
      "name": "test",
      "prompt": "{{{ input }}}\n\nWrite a comprehensive set of unit tests for the selected code. It should setup, run tests that check for correctness including important edge cases, and teardown. Ensure that the tests are complete and sophisticated. Give the tests just as chat output, don't edit any file.",
      "description": "Write unit tests for highlighted code"
    }
  ],
  "tabAutocompleteModel": {
    "title": "Codestral",
    "provider": "mistral",
    "model": "codestral-latest",
    "apiKey": "PnhawsLZIHCludnW7V4NS5FJ4SjI5ViN"
  },
  "embeddingsProvider": {
    "provider": "free-trial"
  },
  "reranker": {
    "name": "free-trial"
  }
}
```

### Description

The file indexer throws an error.

### To reproduce

Open a project and the Continue panel and hit the little indicator button at the bottom.

### Log output

```Shell
[Extension Host] error when indexing:  TypeError: Table and inner RecordBatch schemas must be equivalent.
    at new Table (c:\Users\dsieradski\.vscode\extensions\continue.continue-0.8.46-win32-x64\out\extension.js:136754:19)
    at makeArrowTable (c:\Users\dsieradski\.vscode\extensions\continue.continue-0.8.46-win32-x64\out\extension.js:143111:16)
    at LocalTable.add (c:\Users\dsieradski\.vscode\extensions\continue.continue-0.8.46-win32-x64\out\extension.js:153293:44)
    at LanceDbIndex.update (c:\Users\dsieradski\.vscode\extensions\continue.continue-0.8.46-win32-x64\out\extension.js:360745:15)
    at CodebaseIndexer.indexFiles (c:\Users\dsieradski\.vscode\extensions\continue.continue-0.8.46-win32-x64\out\extension.js:428989:37)
    at CodebaseIndexer.refresh (c:\Users\dsieradski\.vscode\extensions\continue.continue-0.8.46-win32-x64\out\extension.js:428856:30)
    at Core.refreshCodebaseIndex (c:\Users\dsieradski\.vscode\extensions\continue.continue-0.8.46-win32-x64\out\extension.js:429842:26)
    at c:\Users\dsieradski\.vscode\extensions\continue.continue-0.8.46-win32-x64\out\extension.js:429799:11
    at c:\Users\dsieradski\.vscode\extensions\continue.continue-0.8.46-win32-x64\out\extension.js:469223:20
    at i.value (c:\Users\dsieradski\.vscode\extensions\continue.continue-0.8.46-win32-x64\out\extension.js:435971:32)
```

**Comment by selfagency** [comment]
Ok, I just looked at the directions and setup VoyageAI as an embedding provider and now it works without an error. So you probably need to pop-up a window that says "No embedding provider has been configured" when someone clicks on the button.

**Comment by Patrick-Erichsen** [comment]
I think this error might have been unrelated to your configuration of VoyageAI. Should be patched in https://github.com/continuedev/continue/pull/2020.

**Comment by selfagency** [comment]
@Patrick-Erichsen You're correct because now with it configured, I'm still getting the error.

**Comment by IngAjVillalon** [comment]
I'm now encountering the same error, which is strange since the file indexing worked perfectly yesterday with a green status.
However, I'm now receiving the error: `TypeError: Table and inner RecordBatch schemas must be equivalent`.
I've tried everything, reloading the window, uninstalling and reinstalling the extension, but the error persists.
**Any suggestions to fix the indexing would be greatly appreciated.** 
### Thank you for your excellent work; it helps many developers improve their applications.

**Comment by anasalaqeel** [comment]
Same here

**Comment by hanxu98121** [comment]
I got same error as well.

**Comment by aimer63** [comment]
I got the same error. 
It started after updating to version 0.8.46, no problem before.

**Comment by a1eksandrk** [comment]
Same error on the old project. No problems on the new one

**Comment by freakontrol** [comment]
Same for me

**Comment by klcd** [comment]
Same problem here

**Comment by kyleboddy** [comment]
I've seen this problem intermittently as well.

**Comment by thyeun** [comment]
Same here but it appear after a hours of using it, meaning that after finish up all the download and installation (follow exactly the step by step from the extension page), this error, will not happen immediately, it will appear after a hour or after the system reboot.

**Comment by PhysicistLeon** [comment]
What has helped so far is deleting the contents of `~/.continue/` and rolling back to v0.8.45

**Comment by marcov-dart** [comment]
> What has helped so far is deleting the contents of `~/.continue/` and rolling back to v0.8.45

Well, I did that... And then it indexed all my files and it worked for ~10 minutes or so. And then:

Error updating the codeSnippets index: Error: SQLITE_CONSTRAINT: UNIQUE constraint failed: code_snippets.path, code_snippets.cacheKey, code_snippets.content, code_snippets.title, code_snippets.startLine, code_snippets.endLine

**Comment by thyeun** [comment]
@marcov-dart You should not deleted the all file inside ```~/.continue/```, just as i mentioned, only deleted the folder inside ```~/.continue/index/lancedb/```, reason is, several ```config``` file still need it to be the function to do the re-index step, if you deleted it, meaning that the whole function to do the re-index will be block. That what i see from the structure of ```~/.continue``` folder.

To me, after deleted the folder and re-index, it been a day no more error appear. But i dont think it is the solution.

**Comment by adrianodias8** [comment]
> @marcov-dart You should not deleted the all file inside `~/.continue/`, just as i mentioned, only deleted the folder inside `~/.continue/index/lancedb/`, reason is, several `config` file still need it to be the function to do the re-index step, if you deleted it, meaning that the whole function to do the re-index will be block. That what i see from the structure of `~/.continue` folder.
> 
> To me, after deleted the folder and re-index, it been a day no more error appear. But i dont think it is the solution.

In my case i did as you said (went to `~/.continue/index/lancedb/` and delete folders only related to a specific project name)

next time i open vscode it started indexing again but then i got the same exact issue 

`Error updating the codeSnippets index: Error: SQLITE_CONSTRAINT: UNIQUE constraint failed: code_snippets.path, code_snippets.cacheKey, code_snippets.content, code_snippets.title, code_snippets.startLine, code_snippets.endLine`

**Comment by thyeun** [comment]
@marcov-dart i can confirm that your error is missing file from the below

1. ```autocompleteCache.sqlite```
2. ```docs.sqlite```
3. ```globalContext.json```
4. ```index.sqlite```

**Comment by marcov-dart** [comment]
You may be right, but all the files you mention exist in my .continue folder. They seem to have been recreated. And indexing started at the top, reindexing everything. It managed to finished ok. And then 10 minutes or so of working on my code suddenly the SQLITE_CONSTRAINT: UNIQUE constraint failed: code_snippets.path, code_snippets.cacheKey, code_snippets.content, code_snippets.title, code_snippets.startLine, code_snippets.endLine.

**Comment by marcov-dart** [comment]
Also, why can't I delete everything?
What would be the correct procedure to restart using continue entirely from scratch clearing all history and everything if not to delete this .continue folder?

**Comment by edison** [comment]
Same problem here since I started using in last days. I just started to use Continue, was pretty excited and then became sad that this is not working.

> Indexing error: "TypeError: Table and inner RecordBatch schemas must be equivalent."

I'm happy to see that it's a recent issue and maybe someone find a solution.

**Comment by fry69** [comment]
Please delete only `~/.continue/index` completely and leave all other parts inside `~/.continue` alone.

Do not delete only `~/.continue/index/lancedb/` or single files there, other parts in `~/.continue/index` depend on this and get out of sync.

**Comment by marcov-dart** [comment]
Ok, then it works. For about 5 minutes. 
When I start working, saving files etc then soon after it is broken again. Using continue as it stands is more bother than it is worth. I don't want to be fighting tools I depend on so I sadly deinstalled continue.

**Comment by thyeun** [comment]
Not sure what happen to your system @marcov-dart, but from my first post till now, the error no more appear. 

Possible to remove all and reinstall back from step one?

**Comment by marcov-dart** [comment]
Already done that 3 times. It will index everything which takes a while but seems to work fine. After that I do manual reindex which finishes quickly and also no problem. I can add files, modify files and remove files and reindex with no problems . Also seems to work. And then I start working and after a bit the error pops up and it stops working.

**Comment by fry69** [comment]
@marcov-dart Which editor and which Continue version do you use?

**Comment by marcov-dart** [comment]
VsCode on macOS with continue v0.8.46 (latest release). 
I also tried previous and the current pre-release version, but no dice.

**Comment by fry69** [comment]
Switching between version can create problems, esp. when going backwards from a Pre-Release. I'd recommend to wipe out the `index` folder in that case.

Also I'd recommend to stick with Pre-Releases for now, if possible, they include many fixes for the indexer. If possible also check if your code base includes files that might be problematic for the indexer (e.g. binary files, files with very long lines). Most of those files should get excluded automatically, but edge cases exist.

You can check the default ignore/exclusion patterns here -> https://github.com/continuedev/continue/blob/main/core/indexing/ignore.ts

**Comment by marcov-dart** [comment]
Well, the pre-release also doesn't work. But it is ok. I removed continue and will no longer be using it.

**Comment by thyeun** [comment]
There are two scenario.

1. If you open a folder, it will automatically index all the file/sub-folder into ~/.continue/index/lancedb/homexxxxxxmainvectordbOllamaEmbeddingsProvidernomic-embed-text.lance (the xxxxxx is the path folder you want to open) and at ```continue``` icon, you will see it starting index from 0% to 100%, if successful index, you will see a green button, if fail, you will get a red button (with error messages). while if it start indexing, pls dont interrrupt it, if not you will fail to index, and you will get red button (with error messages).
2. If you open single file, it will not doing any index, and at ```continue``` icon, you will see a gray button.

That what happen to my system.

**Comment by marcov-dart** [comment]
@thyeun 

Yeah it indexes from 0 to 100% then it reports it is done with a green icon. It has a created a whole bunch of folders inside the lancedb folder. (I have a few dozen folders in my workspace). It takes a while and I don't interrupt it.

But it just does not keep working. It will just fail after a while and suddenly show the red icon. I assume some actions like saving a file triggers the reindexing or maybe it is on a timer I am not sure.

And the error will not clear by itself. Restarting vscode or restarting the whole computer has no effect. The database has gotten corrupted or something. I'll have to reset everything and wait again for all the embeddings to be calculated. 

As it is continue is unworkable at least for me on my system.

**Comment by fry69** [comment]
Some remote index services can be problematic. I have seen a few reports of Voyager being involved with error or corrupt databases (in the end).

If possible I'd recommend to stick to the default `transformers.js` based embeddings model (included in VSCode, just remove all `embeddingsProivder` and `reranker` blocks from your `config.json`), or run a local embeddings model via Ollama (embeddings models are not LLMs, they don't need much compute and should run fine on CPU), see here -> https://docs.continue.dev/features/codebase-embeddings#ollama

**Comment by marcov-dart** [comment]
I *AM* using the default transformers.js

**Comment by fry69** [comment]
> I _AM_ using the default transformers.js

Interesting. Do you any other special setup? Remote SSH or other (dev-)container based things?

**Comment by marcov-dart** [comment]
Nope. I have nothing special like that. Its all local filesystem on macos.

**Comment by fry69** [comment]
> Nope. I have nothing special like that. Its all local filesystem on macos.

Thank you for the valuable information.

**Comment by thyeun** [comment]
i only can reproduce if i manually force to re-index, after i make change on a file (save it also), than i will getting the same message as yours.

But if i not doing manually force to re-index, and finish up what ever things be change on the file (save it also) than after closing vs code, the next round after i open up vs code and i can see the ```continue``` icon will start automatically indexing. It will index successfully.  

That what i can try and reproduce the error. It might different from others, or someone can try to reproduce?

**Comment by marcov-dart** [comment]
Ok, I will try one more time. I have cleared everything and re-installed continue. I have not modified any settings and it immediately started the indexing. It says computing embeddings for 100 files with a progress bar. I will go get dinner and let it do its thing. I will not press the thing for manual reindexing. And see if it holds. I will report back.

**Comment by fry69** [comment]
FYI: Really huge repositories (like e.g. the Linux kernel with ~90k files and 5GB total size) will break the indexer currently btw. Medium sized repositories (<10k files and <1GB size) should be fine (just a rough guess, not exactly clear where the breaking point lies and there are already fixes in discussion to get rid of this problem).

**Comment by thyeun** [comment]
@fry69 Yes, i try to reproduce with 4.8GB folder, than i will not possible to finish up the indexing, and will get fail message, after restart vs code (meaning that close and re-open back).

Kindly to double confirm here, if indexing fail, before close up vs code, for my system, i still can see the indexing button turn to green, but after close up and re-open back vs code, than only the button turn red with error message.

**Comment by edison** [comment]
Never seen the indexing error again after deleting all files inside `~/.continue/index/lancedb/`.

P.S.: This is a great tool, can't be more grateful to people who develop and support this project. Thank you guys!

**Comment by fry69** [comment]
It is also possible to disable indexing for a specific workspace folder / project. Create a file called `.continuerc.json` in the workspace root and put
```json
{
	"disableIndexing": true
}
```
into it.

**Comment by marcov-dart** [comment]
I do not think the size of my stuff is the problem. I suspect the problem is due to unclean shutdown of indexing task in my case when switching workspaces for instance while the indexing task is running. I had the SQLITE_BUSY: database is locked when switching workspaces one time.

**Comment by thyeun** [comment]
@fry69 continuerc.json already inside ~/.continue/, or you mean for a folder that existed 5GB will need to create a file continuerc.json in this folder to avoid indexing?

Is this a solution or a temporary fix?

**Comment by sestinj** [comment]
Thanks everyone for sharing feedback. A fix for this is on the way, as well as in general better tools to manage a corrupt state when necessary. Look out for a pre-release later today, and going forward our indexing is much more thoroughly tested, so you can expect less "fighting" of the tool. If worst comes to worst, "disableIndexing" will remain an option

**Comment by ameer-clara** [comment]
Great tool, awesome work :) 

I installed it yesterday and see this error. How do you install the pre-release version? What is the official 'fix' from the dev team for the current production version, if index is not disabled.

**Comment by plashenkov** [comment]
Can confirm that latest pre-release version works fine for me now.

Before that nothing worked - manually deleting index folder, etc. Given that I used default settings for embeddings (no configuration in config.json), and my codebase was extremely small - just index.html and index.js for testing purposes.

Thank you for this update!

**Comment by kyleboddy** [comment]
In a folder with a lot of files, I used to get the above issue, but not with my other smaller project.

When I switch to pre-release, I get this problem in the "Extension Host" dropdown of the Output pane:

> 2024-08-22 01:05:44.539 [error] Error: SQLITE_CONSTRAINT: UNIQUE constraint failed: tag_catalog.dir, tag_catalog.branch, tag_catalog.artifactId, tag_catalog.path, tag_catalog.cacheKey
> 2024-08-22 01:05:44.649 [error] Error: SQLITE_CONSTRAINT: UNIQUE constraint failed: tag_catalog.dir, tag_catalog.branch, tag_catalog.artifactId, tag_catalog.path, tag_catalog.cacheKey
> 2024-08-22 01:05:44.744 [error] Error: SQLITE_CONSTRAINT: UNIQUE constraint failed: tag_catalog.dir, tag_catalog.branch, tag_catalog.artifactId, tag_catalog.path, tag_catalog.cacheKey

![image](https://github.com/user-attachments/assets/71b0330a-9bcb-4b4c-a040-ccc7a3321d75)

And this issue when hovering over the error marker in my other project.

![image](https://github.com/user-attachments/assets/e1ba5ec0-5679-41f3-847f-a42c36faecd1)

These are both remote-ssh folders.

The errors persist when I switch back to the Release candidate version. I uninstall and reinstall, and it works for my smaller project (Python) but still does not work for the larger project (PHP).

**Comment by thyeun** [comment]
@kyleboddy Yes, if your folder larger than 5GB (a file size with more than 1mb) than this error will become very obvious.

Do you try out the pre-release version?

**Comment by kyleboddy** [comment]
Yes, I did and ran into the SQLITE_CONSTRAINT errors above.

> When I switch to pre-release, I get this problem in the "Extension Host" dropdown of the Output pane:

**Comment by thyeun** [comment]
@kyleboddy 100mb you have to wait till the indexing process finish from 0% to 100%, it might need sometime, depend also your file size. But if successful indexing to 100%, your case should be no problem.

**Comment by marcov-dart** [comment]
Was there a new pre-release for the vscode extension? I am not sure. I didn't notice it updating though.

In any case I tried v0.9.198 (pre-release) and this time it wasn't even able to finish to 100% after resetting everything and just leaving it alone without it failing.

**Comment by ameer-clara** [comment]
After taking installing the pre-release version, there is a new error when using ```@codebase```, 

Error getting context items from codebase: TypeError: db.search is not a function

download open-webui and try ```@codebase``` to repro. Not ready for primetime just yet, invest in qa :)

**Comment by anotherbazeinthewall** [comment]
Experienced same issue on Mac OS (some workspaces but not all), switching to latest pre-release version resolved. Thank you so much, guys!

**Comment by marcov-dart** [comment]
Vscode extension pre-release v0.9.199 does NOT fix the problems on Mac OS for me.

1) I deleted the whole index folder.
2) It will start indexing and does complete with "index up to date"
3) The index folder is 293 Mb in size afterwards
4) I cannot close vscode and reopen it. It will report "Indexing error - click to retry" and when hovering over it then it says "TypeError: Cannot read properties of undefined (reading 'id')"
5) I cannot do a manual indexing after it finished and says "index up to date" either. It will start saying "indexing 3 files" and bug out.

I don't know what is going on or why it is so hard to get this right, but there you go.

**Comment by marcov-dart** [comment]
Yeah, so it was ok for about an hour and then I switched to another workspace. The indexing at the time was not running or at least not showing to be running. I checked before switching. It started reindexing after switching to the new workspace and then failed with "TypeError: Cannot read properties of undefined (reading 'id')".

**Comment by thyeun** [comment]
@marcov-dart Are you using pre-release?

**Comment by marcov-dart** [comment]
Was Specified:
Vscode extension **pre-release v0.9.199** does NOT fix the problems on Mac OS for me.

**Comment by breynolds3** [comment]
I did some debugging on this. the schema is in fact different when trying to store the files, but I have no idea why.  Attaching the json export of the schema of the db vs the requested schema.  It's an intermittent problem.
[batch.schema.json](https://github.com/user-attachments/files/16741750/batch.schema.json)
[schema.json](https://github.com/user-attachments/files/16741751/schema.json)
```
--- schema.json	2024-08-23 09:43:09.039580400 -0400
+++ batch.schema.json	2024-08-23 09:43:43.411651000 -0400
@@ -41,7 +41,7 @@
             "type": {
                 "precision": 2
             },
-            "nullable": false,
+            "nullable": true,
             "metadata": {}
         },
         {
@@ -49,13 +49,13 @@
             "type": {
                 "precision": 2
             },
-            "nullable": false,
+            "nullable": true,
             "metadata": {}
         },
         {
             "name": "contents",
             "type": {},
-            "nullable": false,
+            "nullable": true,
             "metadata": {}
         }
     ],
```

**Comment by sestinj** [comment]
This should now be completely resolved. If anyone is still experiencing the problem though, please re-open the issue and let me know!

**Comment by casperdcl** [comment]
fixed for me in v0.9.200 (WSL2)

**Comment by marcov-dart** [comment]
Same, no problems here anymore either (for the last 4 hours). Is looking good. v0.9.200 as well.
Thanks!

**Comment by marcov-dart** [comment]
Unfortunately I spoke to soon...

I opened up a 2nd workspace. The 2nd workspace started to do some indexing but then seems to be blocked by the other. And then after closing vscode and starting it back up now "SQLITE_BUSY: database is locked"

**Comment by MatsPro** [comment]
Fix with the new version worked for me too. I'm running local with only one workspace however.

**Comment by thyeun** [comment]
I'm still using 0.8.47 version (linux), dont hv any issue, can auto-indexing, re-indexing and etc for more than 2 to 3 days.
